### ISSUE TITLE
Implement a `git reword` command

### ISSUE THREAD

**Issue by bcongdon** [arxanas/git-branchless#179]
As discussed in #48.

**Comment by epage** [comment]
I'm curious, do you expect `git reword` to be tied to `git branchless init` or would it be like https://github.com/arxanas/git-branchless/issues/183?  I'd like
- `git revise -ie` like behavior
- Branches to be preserved
- Able to be used independent of `git branchless`.

Trying to decide if its worth making my own or if I can somehow install yours without the other parts (or just not run the other parts)

**Comment by arxanas** [comment]
I think it would be reasonable to make `git reword` work without `git branchless init`/hooks.

**Comment by bcongdon** [comment]
> If you reword multiple commits at once, then it opens an editor with all of their commit messages separated by some marker.

That sounds like a useful feature. 👍 I was thinking that even in the single-commit case, there should be support for dropping into the `$EDITOR`. e.g. `git reword -m "new message"` is legal, but the bare `git reword` would drop you into the commit message editor similar to `git commit --amend`. Maybe the single-commit case is just a special case of multi-commit reword.

**Comment by arxanas** [comment]
I wonder what the behavior should be for `git reword -m "new message" 'revset_expr_returning_multiple_commits()'`?

For the future implementor, `dialoguer` has a function to spawn an editor, which might be useful:https://docs.rs/dialoguer/0.9.0/dialoguer/struct.Editor.html

**Comment by claytonrcarter** [comment]
Perhaps something like:

- `git reword` - open `$EDITOR`, amend `HEAD`
- `git reword <commit>` - open `$EDITOR`, amend `<commit>`
- `git reword <multiple commits>` - open `$EDITOR` for bulk editing of `<multiple commits>`
- `git reword -m <message>` - use `<message>` to amend `HEAD`
- `git reword <commit> -m <message>` - use `<message>` to amend `<commit>`
- `git reword <multiple commits> -m <message>` - use same `<message>` to amend all `<multiple commits>`, possibly printing a warning to stderr or something like that

Personally, I could see myself using all of these. The last is the oddest, of course, but could be useful for something like `git reword <buncha commits on branch> -m 'fixup!' && git rebase -i --autosquash main`

**Comment by claytonrcarter** [comment]
For those playing along at home, `git reword` was recently merged into `master` via #304. To the best of my knowledge, it handles everything discussed here *except* bulk edit mode a la `git revise -ie`. I would love hear your thoughts and experiences from using it in it's present form.

Bulk edit mode is next up, but I'm going to try *really* hard to take a short break before digging into it. If you have any ideas or suggestions for that feature, please share them. FWIW, my current plan is to stick close to the format used by `git revise -ie` so that folks that use both will feel comfortable switching back and forth, and to avoid any unnecessary wheel reinvention. Something like this:

```
++ reword abc123
This is the commit message for abc123.

++ reword def456
This is the commit message for def456.

# Rewording 2 commits. 
# ... some helpful, descriptive message here ...
```

**Comment by arxanas** [comment]
Since we have support for a fairly functional `git reword` command, I'm going to close this issue. We can discuss further improvements/features in Discussions or on Discord. @claytonrcarter thanks again 😄
